kramer/src/main/kotlin/kramer/MavenRepo.kt

```diff
@@ -0,0 +1,331 @@
+package kramer
+
+import com.github.ajalt.clikt.core.CliktCommand
+import com.github.ajalt.clikt.core.ProgramResult
+import com.github.ajalt.clikt.core.requireObject
+import com.github.ajalt.clikt.parameters.options.default
+import com.github.ajalt.clikt.parameters.options.defaultLazy
+import com.github.ajalt.clikt.parameters.options.option
+import com.github.ajalt.clikt.parameters.options.required
+import com.github.ajalt.clikt.parameters.types.int
+import com.github.ajalt.clikt.parameters.types.path
+import com.squareup.tools.maven.resolution.ArtifactResolver
+import com.squareup.tools.maven.resolution.FetchStatus
+import com.squareup.tools.maven.resolution.FetchStatus.RepositoryFetchStatus.SUCCESSFUL
+import com.squareup.tools.maven.resolution.FetchStatus.RepositoryFetchStatus.SUCCESSFUL.FOUND_IN_CACHE
+import com.squareup.tools.maven.resolution.Repositories
+import com.squareup.tools.maven.resolution.ResolvedArtifact
+import java.io.IOException
+import java.net.URI
+import java.nio.file.FileSystem
+import java.nio.file.FileSystemAlreadyExistsException
+import java.nio.file.FileSystems
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.Collections
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+import javax.xml.parsers.SAXParserFactory
+import kotlin.system.measureTimeMillis
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.filterNot
+import kotlinx.coroutines.flow.flatMapMerge
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.runBlocking
+import org.xml.sax.Attributes
+import org.xml.sax.SAXException
+import org.xml.sax.helpers.DefaultHandler
+
+class MavenRepo(fs: FileSystem = FileSystems.getDefault()) : CliktCommand(name = "gen-maven-repo") {
+  private val workspace: Path by option(
+    "--workspace",
+    help = "Path to the workspace to be generated."
+  )
+    .path(canBeFile = false, canBeSymlink = false, canBeDir = true)
+    .default(fs.getPath("${System.getProperties()["java.io.tmpdir"]}", "bazel/maven"))
+
+  /** Name of the workspace - nearly always the nearest directory name of the workspace path */
+  private val workspaceName: String by option("--workspace-name")
+    .defaultLazy { workspace.fileName.toString() }
+
+  private val configFile by option("--configuration").path(mustExist = true).required()
+
+  private val threadCount: Int by option("--threads").int().default(1)
+
+  internal val kontext by requireObject<Kontext>()
```

Comment:
Help me out here -- does this mean Kontext must be defined, or does it create one?

Reply:
This is the special API klikt uses this requireObject means that klikt will initiate this, passing it down the "inheritance" (in the binary/command-subbinary sense) chain. Kontext is defined in Kramer and since MavenRepo is installed into Kramer as a sub-command, Kramer becomes its parent binary. So this requireObject is a fancy type-safe map lookup into its "parent". 

It's gross, but talk to the Clikt folks. I think it's like this because the other option was to make actual real type inheritance happen, and in really complex apps, where you have a various type-hierarchy bits in play, that would suck.